Index prefixes of tokens

levjj opened the pull request:
The current version of reds uses metaphonics and stemming but does not index prefixes, so when indexing "LearnBoost", you can find the id when querying "Lirnbast" but you cannot find it when searching for "LearnBoos".

This patch also indexes prefixes but I guess you probably want an option that turns this feature on or off.  This does not affect the performance of queries but makes indexing and removing somewhat slower.

stockholmux commented:
Thanks for the PR. I took some time to play around with your implementation. I think you have a point with the problem presented and I like how you approached it.

I've got a few issues with it, however:
1. This greatly increases the time to index / remove but also the space required in Redis for the index. Since this is based on an in-memory datastore, that is a sensitive area. Not everyone will want that.
2. In `reds.js` on 318 you've got a magic number of 10 - that is the score 'bonus' for having the exact match? Might be nice to make that configurable.
3. This definitely should be optional or, better yet, a separate module that works with reds. Since you are just modifying a couple of functions, perhaps you can make a module that accepts a reds search object and overrides the `Search.prototype.remove` and `Search.prototype.index` functions. If there is anything else that I could do to make reds more pluggable, I'm willing to hear it (I think pluggablity will be the future of this module, really since people sometimes bring very specific needs as PRs).

plus- commented:
This fixes the previous fuzzy search issue mentioned once before then

https://github.com/tj/reds/issues/38

levjj commented:
Thanks for the comments.  I just updated the PR in response to your feedback.

1. If the default length of words metaphones (without vowels etc.) is about 5 letters, then prefix matching would increase the memory from 2 set members per word to 6 set members per word.  So, I would estimate that time for indexing, removing and space all increase by a factor of 3.  Whether this is acceptable depends on the application.

2. Yes, it would be good to rank exact matches higher than matches that just happen to be the prefix of a longer word.  However, choosing the right weight is difficult, so it should be made configurable. The current version of reds does not have a way to pass in options, so I added this as additional (optional) argument to `createSearch` with some default values (and documentation).  By the way, with options it is also possible to enable/disable stemming or metaphone mapping as requested in another Github issue.

3. This should definitely be an optional feature. By passing in options to `createSearch`, it is possible to enable this feature on demand while reds does not index prefixes by default. Doing this as external module would not be ideal. Of course, you can always wrap the reds libary or monkey-patch `Search` objects but this makes it difficult to combine it with other reds plugins. I would suggest to either make this a part of reds or to introduce a clean way of writing reds plugins.

stockholmux commented:
#45 - discussion on plugins.

stockholmux commented:
You can now implement this in a `nlpProcess` function (see #50)
